includes/core/class-mode.php

```diff
@@ -0,0 +1,172 @@
+
+	/**
+	 * Holds the dev mode flag.
+	 *
+	 * @var bool
+	 */
+	private $dev_mode;
+
+	/**
+	 * Holds the gateway class for settings.
+	 *
+	 * @var WC_Payment_Gateway_WCPay
+	 */
+	private $gateway;
+
+	/**
+	 * Environment types, which are used to automatically enter dev mode.
+	 *
+	 * @see wp_get_environment_type()
+	 * @see https://developer.wordpress.org/reference/functions/wp_get_environment_type/#description
+	 */
+	const DEV_MODE_ENVIRONMENTS = [
+		'development',
+		'staging',
+	];
+
+	/**
+	 * Stores the gateway for later retrieval of options.
+	 *
+	 * @param WC_Payment_Gateway_WCPay $gateway The active gateway.
+	 */
+	public function __construct( WC_Payment_Gateway_WCPay $gateway ) {
+		$this->gateway = $gateway;
+	}
+
+	/**
+	 * Initializes the working mode of WooCommerce Payments.
+	 */
+	private function init() {
+		$dev_mode = (
+			// Plugin-specific dev mode.
+			$this->is_wcpay_dev_mode_defined()
+
+			// WordPress Dev Environment.
+			|| in_array( $this->get_wp_environment_type(), self::DEV_MODE_ENVIRONMENTS, true )
+		);
+
+		/**
+		 * Allows WooCommerce to enter dev mode.
+		 *
+		 * @see https://woocommerce.com/document/payments/testing/dev-mode/
+		 * @param bool $dev_mode The pre-determined dev mode.
+		 */
+		$this->dev_mode = (bool) apply_filters( 'wcpay_dev_mode', $dev_mode );
+
+		$test_mode_setting = 'yes' === $this->gateway->get_option( 'test_mode' );
+		$test_mode         = $this->dev_mode || $test_mode_setting;
+
+		/**
+		 * Allows WooCommerce to enter test mode.
+		 *
+		 * @see https://woocommerce.com/document/payments/testing/
+		 * @param bool $test_mode The pre-determined test mode.
+		 */
+		$this->test_mode = (bool) apply_filters( 'wcpay_test_mode', $test_mode );
+	}
+
+	/**
+	 * Checks whether WCPay is in the given mode.
+	 *
+	 * @param  string $property The mode to check for. Either `live`, `test`, or `dev`.
+	 * @return bool If the propery is toggled.
+	 *
+	 * @throws Exception In case the class has not been initialized yet.
+	 */
+	public function __get( $property ) {
+		if ( ! isset( $this->dev_mode ) || ! isset( $this->test_mode ) ) {
+			if ( isset( $this->gateway ) && ! empty( $this->gateway->settings ) ) {
+				$this->init();
+			} else {
+				throw new Exception( 'WooCommerce Payments\' working mode is not initialized yet. Wait for the `init` action.' );
+			}
+		}
+
+		if ( 'live' === $property ) {
+			return ! $this->test_mode && ! $this->dev_mode;
+		} else {
+			$prop = $property . '_mode';
+			if ( property_exists( $this, $prop ) ) {
+				return $this->$prop;
+			}
+		}
+	}
+
+	/**
+	 * Forces a switch of the current mode.
+	 * Only recommended when testing, use filters otherwise.
+	 *
+	 * @param string $name The name of the mode to enter.
+	 * @param array  $args All args for the method (not used).
+	 * @return void
+	 */
+	public function __call( $name, $args ) {
+		// Only simulate the possible modes.
+		if ( ! in_array( $name, [ 'live', 'test', 'dev' ], true ) ) {
+			return;
```

Comment:
I would throw `BadMethodCallException` here to catch issues with methods that we don't support here to prevent possible issues with calling random functions that do not exist

Reply:
Not applicable after the switch to normal methods in https://github.com/Automattic/woocommerce-payments/pull/4986/commits/896eaa3f9aac54d505f17ac81b0d71d48a9b9534